File: src/difi/partitions.py
from typing import Optional

import pyarrow as pa
import pyarrow.compute as pc
import quivr as qv
from quivr.validators import and_, ge, le


class Partitions(qv.Table):
    #: Partition ID - Represents a division of the data whether by a linking window
    #: or some other criteria.
    id = qv.LargeStringColumn()
    #: Start night of the partition.
    start_night = qv.Int64Column()
    #: End night of the partition (inclusive).
    end_night = qv.Int64Column()

    @classmethod
    def create_single(cls, nights: pa.Array) -> "Partitions":
        """
        Create a single partition that spans the given nights.

        Parameters
        ----------
        nights : pa.Array
            Array of nights to span (does not need to be sorted or unique).

        Returns
        -------
        paritions : Partitions
            Single partition spanning the given nights.
        """
        return cls.from_kwargs(
            id=["0"],
            start_night=[pc.min(nights).as_py()],
            end_night=[pc.max(nights).as_py()],
        )

    @classmethod
    def create_linking_windows(
        cls,
        nights: pa.Array,
        detection_window: Optional[int] = None,
        min_nights: Optional[int] = None,
        sliding: bool = False,
    ) -> "Partitions":
        """
        Create partitions that represent linking windows of observations. If detection_window is
        None, then the entire range of nights is used. If detection_window is larger than the range
        of the observations, then the entire range of nights is used. If detection_window is smaller
        than the range of the observations, then one of two things can happen. If sliding is True,
        then the windows will shift by one night at a time with a minimum length of min_nights. If
        sliding is False, then the windows will be non-overlapping with a minimum length of detection_window.

        Parameters
        ----------
        nights : pa.Array
            Array of nights on which observations occur.
        detection_window : int, optional
            The number of nights of observations within a single window. If None, then the entire range
            of nights is used as a single window (both min_nights and sliding are ignored).
        min_nights : int, optional
            Minimum length of a detection window measured from the earliest night (only applies when
            sliding is True). If the detection window is set to 15 but min_nights is 3
            then the first window will be 3 nights long and the second window will be 4 nights long, etc...
            Once the detection_window length has been reached then all windows will be of length detection_window.
        sliding : bool, optional
            If True, then the windows will be sliding windows of length detection_window. If False, then the
            windows will be non-overlapping windows of length detection window.

        Returns
        -------
        windows : Partitions
            Table detailing the start and end night of each window.
        """
        # Calculate the unique number of nights
        min_night = pc.min(nights).as_py()
        max_night = pc.max(nights).as_py()

        # If the detection window is not specified, then use the entire
        # range of nights
        if detection_window is None:
            partitions = Partitions.create_single(nights)
            return partitions

        if detection_window >= (max_night - min_night):
            detection_window = max_night - min_night + 1

        if min_nights is None:
            min_nights = detection_window

        if detection_window < min_nights:
            raise ValueError("Detection window must be greater than or equal to min_nights.")

        partitions = cls.empty()

        if sliding:

            i = 0
            start_night = min_night
            end_night = start_night + min_nights - 1
            while True:
                if end_night > max_night:
                    break

                partitions_i = Partitions.from_kwargs(
                    id=[f"{i}"], start_night=[start_night], end_night=[end_night]
                )
                partitions = qv.concatenate([partitions, partitions_i])

                i += 1
                end_night += 1
                if end_night - detection_window == start_night:
                    start_night += 1

        else:
            for i, start_night in enumerate(range(min_night, max_night, detection_window)):

                # End night is inclusive
                end_night = start_night + detection_window - 1
                if end_night > max_night:
                    end_night = max_night

                partitions_i = Partitions.from_kwargs(
                    id=[f"{i}"], start_night=[start_night], end_night=[end_night]
                )
                partitions = qv.concatenate([partitions, partitions_i])

        return partitions


class PartitionSummary(qv.Table):
    #: Partition ID - Represents a division of the data whether by a linking window
    #: or some other criteria.
    id = qv.LargeStringColumn()
    #: Start night of the partition.
    start_night = qv.Int64Column()
    #: End night of the partition (inclusive).
    end_night = qv.Int64Column()
    #: Number of observations in the partition.
    observations = qv.Int64Column()
    #: Number of unique objects that are deemed findable in the partition.
    findable = qv.Int64Column(nullable=True)
    #: Number of unique objects that were found in the partition.
    found = qv.Int64Column(nullable=True)
    #: The completeness of the partition (found / findable).
    completeness = qv.Float64Column(nullable=True, validator=and_(ge(0), le(1)))
    #: Number of pure linkages of known objects.
    pure_known = qv.Int64Column(nullable=True)
    #: Number of pure linkages of unassociated observations.
    pure_unknown = qv.Int64Column(nullable=True)
    #: Number of contaminated linkages.
    contaminated = qv.Int64Column(nullable=True)
    #: Number of mixed linkages.
    mixed = qv.Int64Column(nullable=True)

File: src/partitions.rs
//! Partition creation and summary.
//!
//! Partitions divide observations into time windows for analysis.
//! Each partition is defined by a start and end night (inclusive).

use crate::error::{Error, Result};

/// A partition defining a time window of observations.
#[derive(Debug, Clone)]
pub struct Partition {
    pub id: u64,
    pub start_night: i64,
    pub end_night: i64,
}

/// Summary statistics for a partition.
#[derive(Debug, Clone)]
pub struct PartitionSummary {
    pub id: u64,
    pub start_night: i64,
    pub end_night: i64,
    pub observations: i64,
    pub findable: Option<i64>,
    pub found: Option<i64>,
    pub completeness: Option<f64>,
    pub pure_known: Option<i64>,
    pub pure_unknown: Option<i64>,
    pub contaminated: Option<i64>,
    pub mixed: Option<i64>,
}

/// Create a single partition spanning all given nights.
pub fn create_single(nights: &[i64]) -> Result<Partition> {
    if nights.is_empty() {
        return Err(Error::InvalidInput(
            "Cannot create partition from empty nights".to_string(),
        ));
    }
    let min_night = *nights.iter().min().unwrap();
    let max_night = *nights.iter().max().unwrap();
    Ok(Partition {
        id: 0,
        start_night: min_night,
        end_night: max_night,
    })
}

/// Create non-overlapping or sliding linking windows.
///
/// If `detection_window` is None, returns a single partition spanning all nights.
/// If `sliding` is true, windows slide by one night with a ramp-up from `min_nights`.
/// If `sliding` is false, windows are non-overlapping blocks of `detection_window` nights.
pub fn create_linking_windows(
    nights: &[i64],
    detection_window: Option<i64>,
    min_nights: Option<i64>,
    sliding: bool,
) -> Result<Vec<Partition>> {
    if nights.is_empty() {
        return Err(Error::InvalidInput(
            "Cannot create partitions from empty nights".to_string(),
        ));
    }
    let min_night = *nights.iter().min().unwrap();
    let max_night = *nights.iter().max().unwrap();

    let detection_window = match detection_window {
        None => return Ok(vec![create_single(nights)?]),
        Some(dw) => {
            if dw >= (max_night - min_night + 1) {
                max_night - min_night + 1
            } else {
                dw
            }
        }
    };

    let min_nights = min_nights.unwrap_or(detection_window);
    if detection_window < min_nights {
        return Err(Error::InvalidInput(
            "Detection window must be >= min_nights".to_string(),
        ));
    }

    let mut partitions = Vec::new();

    if sliding {
        let mut i: u64 = 0;
        let mut start_night = min_night;
        let mut end_night = start_night + min_nights - 1;
        loop {
            if end_night > max_night {
                break;
            }
            partitions.push(Partition {
                id: i,
                start_night,
                end_night,
            });
            i += 1;
            end_night += 1;
            if end_night - detection_window == start_night {
                start_night += 1;
            }
        }
    } else {
        let mut i: u64 = 0;
        let mut start = min_night;
        while start <= max_night {
            let end = (start + detection_window - 1).min(max_night);
            partitions.push(Partition {
                id: i,
                start_night: start,
                end_night: end,
            });
            i += 1;
            start += detection_window;
        }
    }

    Ok(partitions)
}

/// Create partition summaries by counting observations per partition.
///
/// Uses a pre-sorted night index for O(log n) lookups per partition.
pub fn create_summaries(
    obs_nights: &[i64],
    partitions: &[Partition],
    night_sorted_indices: &[usize],
) -> Vec<PartitionSummary> {
    partitions
        .iter()
        .map(|p| {
            let lo = night_sorted_indices.partition_point(|&i| obs_nights[i] < p.start_night);
            let hi = night_sorted_indices.partition_point(|&i| obs_nights[i] <= p.end_night);
            let num_obs = (hi - lo) as i64;

            PartitionSummary {
                id: p.id,
                start_night: p.start_night,
                end_night: p.end_night,
                observations: num_obs,
                findable: None,
                found: None,
                completeness: None,
                pure_known: None,
                pure_unknown: None,
                contaminated: None,
                mixed: None,
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_single() {
        let nights = vec![5, 3, 7, 1, 9];
        let p = create_single(&nights).unwrap();
        assert_eq!(p.start_night, 1);
        assert_eq!(p.end_night, 9);
    }

    #[test]
    fn test_create_single_empty() {
        assert!(create_single(&[]).is_err());
    }

    #[test]
    fn test_create_linking_windows_non_overlapping() {
        let nights: Vec<i64> = (0..10).collect();
        let partitions = create_linking_windows(&nights, Some(3), None, false).unwrap();
        assert_eq!(partitions.len(), 4); // [0,2], [3,5], [6,8], [9,9]
        assert_eq!(partitions[0].start_night, 0);
        assert_eq!(partitions[0].end_night, 2);
        assert_eq!(partitions[3].start_night, 9);
        assert_eq!(partitions[3].end_night, 9);
    }

    #[test]
    fn test_create_linking_windows_none() {
        let nights = vec![1, 5, 10];
        let partitions = create_linking_windows(&nights, None, None, false).unwrap();
        assert_eq!(partitions.len(), 1);
        assert_eq!(partitions[0].start_night, 1);
        assert_eq!(partitions[0].end_night, 10);
    }
}
